leakcanary-analyzer/src/test/java/com/squareup/leakcanary/HahaHelperTest.java

```diff
@@ -0,0 +1,239 @@
+package com.squareup.leakcanary;
+
+import com.squareup.haha.perflib.ArrayInstance;
+import com.squareup.haha.perflib.ClassInstance;
+import com.squareup.haha.perflib.ClassObj;
+import com.squareup.haha.perflib.Field;
+import com.squareup.haha.perflib.Snapshot;
+import com.squareup.haha.perflib.Type;
+import com.squareup.haha.perflib.io.HprofBuffer;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.UnsupportedEncodingException;
+import java.util.List;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class HahaHelperTest {
+  public static final int STRING_CLASS_ID = 100;
+  public static final int CHAR_ARRAY_CLASS_ID = 101;
+  public static final int STRING_INSTANCE_ID = 102;
+  public static final int VALUE_ARRAY_INSTANCE_ID = 103;
+
+  public static final int VALUE_ARRAY_LENGTH = 6;
+  public static final int COUNT_VALUE = 5;
+  public static final int OFFSET_VALUE = 1;
+
+  private FakeHprofBuffer buffer;
+  private Snapshot snapshot;
+  private ClassInstance stringInstance;
+
+  @Before
+  public void setUp() {
+    buffer = new FakeHprofBuffer();
+    buffer.setStringsToRead("abcdef");
+
+    snapshot = new Snapshot(buffer);
+    snapshot.setIdSize(4);
```

Comment:
`Snapshot.setIdSize(int)` [sets the size](https://android.googlesource.com/platform/tools/base/+/studio-master-dev/perflib/src/main/java/com/android/tools/perflib/heap/Type.java#17) of `Type.OBJECT` (originally init'd to 0) and initializes `mTypeSizes`.

This method is called by perflib's `HprofParser` [here] (https://android.googlesource.com/platform/tools/base/+/studio-master-dev/perflib/src/main/java/com/android/tools/perflib/heap/HprofParser.java#153) early in the HPROF parsing process according to the [format spec](https://java.net/downloads/heap-snapshot/hprof-binary-format.html).

In test code, not calling `Snapshot.setIdSize(int)` leaves `mTypeSizes` [unallocated](https://android.googlesource.com/platform/tools/base/+/studio-master-dev/perflib/src/main/java/com/android/tools/perflib/heap/Snapshot.java#203) resulting in an NPE when `ClassInstance.getValues()` [attempts to read](https://android.googlesource.com/platform/tools/base/+/studio-master-dev/perflib/src/main/java/com/android/tools/perflib/heap/ClassInstance.java#55) the [instance for the "value" field](https://android.googlesource.com/platform/tools/base/+/studio-master-dev/perflib/src/main/java/com/android/tools/perflib/heap/Snapshot.java#214) in String.java.

Although `Snapshot` does not currently enforce it as a constraint, all instances of `Snapshot` should set the identifier size.